### PR TITLE
MIEngine: Array evaluation and address check

### PR DESCRIPTION
**Pull request by intel-rganesh**
Issue description & fix:

Issue 1
When the address is not an AD7Memory address, we get a null reference exception while assigning puCodeLocationId.
image

Fix
The fix is to add a null reference check for this condition.

Issue 2 
We would like to evaluate array addresses in Memory View. This was not possible without evaluating without an address operator.
image

Fix
For variable information value that looks like [], re-evaluate with address operator.
image
image

With this change we can evaluate an array address.
image

Signed-off-by: intel-rganesh [rakesh.ganesh@intel.com](mailto:rakesh.ganesh@intel.com)

### CODE REVIEW

**Review by WardenGnaw — APPROVED** [review]

File: src/MIDebugEngine/AD7.Impl/AD7Disassembly.cs [hunk 1]
```diff
@@ -34,8 +34,14 @@ public int GetCodeContext(ulong uCodeLocationId, out IDebugCodeContext2 ppCodeCo
...
-            puCodeLocationId = addr.Address;
-            return Constants.S_OK;
+            if (addr != null)
+            {
+                puCodeLocationId = addr.Address;
+                return Constants.S_OK;
+            }
+
+            puCodeLocationId = 0;
+            return Constants.S_FALSE;
```
Comment (gregg-miskelly) [inline src/MIDebugEngine/AD7.Impl/AD7Disassembly.cs:44]: >S_FALSE [](http://example.com/codeflow?start=29&length=7)

Looking at how the other Microsoft debug engines handle this, I would suggest using a failure code here. You can just use `Constants.E_FAIL`.

**Review by gregg-miskelly — CHANGES_REQUESTED** [review]
Looks good aside from those two small issues. Thanks for sending this!

File: src/MIDebugEngine/AD7.Impl/AD7Property.cs [hunk 2]
```diff
@@ -229,11 +229,38 @@ public int GetMemoryContext(out IDebugMemoryContext2 ppMemory)
...
+                // VS on the other hand supports direct array evaluations without address operator
+                // therefore we need to re-evaluate with an address operator
+                //
+                VariableInformation viArray = new VariableInformation("&(" + _variableInformation.FullName() + ")", (VariableInformation)_variableInformation);
+                viArray.SyncEval();
+                if (viArray.Error)
+                {
+                    return AD7_HRESULT.S_GETMEMORYCONTEXT_NO_MEMORY_CONTEXT;
+                }
+                AD7Property propArray = new AD7Property(_engine, viArray);
```
[the remark this answers is not in the document]
Comment (intel-rganesh) [reply src/MIDebugEngine/AD7.Impl/AD7Property.cs]: Thanks, done!

**Review by gregg-miskelly — APPROVED** [review]

**Comment by gregg-miskelly** [comment]
@WardenGnaw have you already investigated what is going on with:
```
D:\a\MIEngine\MIEngine\obj\Release\SSHDebugPS\UI\ContainerPickerDialogWindow.g.cs(50,18): error CS1504: Source file 'UI\ContainerPickerDialogWindow.xaml' could not be opened -- Could not find file. [D:\a\MIEngine\MIEngine\src\SSHDebugPS\SSHDebugPS_cba2tvoo_wpftmp.csproj]
```

**Comment by gregg-miskelly** [comment]
I see. This is what you fixed with https://github.com/microsoft/MIEngine/pull/1425. Any reason not to merge that?

**Comment by WardenGnaw** [comment]
> I see. This is what you fixed with #1425. Any reason not to merge that?

Nope. That PR just got forgotten over the Thanksgiving holidays.

**Pull request merged**